File: examples/customer-support/agent.py
"""Deterministic customer-support example agent for ASE quickstarts."""

from __future__ import annotations

import argparse
import json
import os
import time
from pathlib import Path

AGENT_ID = "customer-support-agent"
SPAN_ID = "customer-support-agent:refund"
TARGET = "https://api.example.com/refunds/ord-001"


def main() -> int:
    """Emit one safe refund flow so quickstarts exercise ASE deterministically."""
    event_path = _event_path(_parse_args())
    event_path.parent.mkdir(parents=True, exist_ok=True)
    event_path.unlink(missing_ok=True)
    _write_events(
        event_path,
        [
            _event("agent_start", agent_id=AGENT_ID, name=AGENT_ID),
            _event(
                "tool_start",
                agent_id=AGENT_ID,
                span_id=SPAN_ID,
                tool_kind="http_api",
                method="POST",
                target=TARGET,
                name="issue_refund",
                data={"order_id": "ord-001"},
            ),
            _event(
                "tool_end",
                agent_id=AGENT_ID,
                span_id=SPAN_ID,
                tool_kind="http_api",
                method="POST",
                target=TARGET,
                status="passed",
                data={"status_code": 200, "order_id": "ord-001"},
            ),
            _event("agent_end", agent_id=AGENT_ID, status="passed"),
        ],
    )
    print("refunded ord-001")
    return 0


def _parse_args() -> argparse.Namespace:
    """Allow explicit event output paths while still honoring ASE's env override."""
    parser = argparse.ArgumentParser()
    parser.add_argument("--events-out", type=Path, default=None)
    return parser.parse_args()


def _event_path(args: argparse.Namespace) -> Path:
    """Prefer explicit CLI output, then ASE runtime output, then local default."""
    if args.events_out is not None:
        return args.events_out
    env_path = os.environ.get("ASE_ADAPTER_EVENT_SOURCE")
    if env_path:
        return Path(env_path)
    return Path(__file__).with_name("events.generated.jsonl")


def _write_events(path: Path, events: list[dict[str, object]]) -> None:
    """Persist events as JSONL without depending on the ASE package import path."""
    with path.open("w", encoding="utf-8") as handle:
        for event in events:
            handle.write(json.dumps(event) + "\n")


def _event(event_type: str, **kwargs: object) -> dict[str, object]:
    """Build adapter events with stable metadata for replay and certification."""
    return {
        "protocol_version": 1,
        "event_type": event_type,
        "event_id": f"{event_type}-{time.time_ns()}",
        "timestamp_ms": time.time() * 1000,
        "metadata": {
            "adapter_name": "customer-support-example",
            "framework": "custom",
            "language": "python",
            "transport": "jsonl-stdio",
        },
        **kwargs,
    }


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/openai-agents-typescript/agent.ts
import { mkdir, writeFile } from 'node:fs/promises';
import { dirname, resolve } from 'node:path';
import { Agent, Usage, run, tool } from '@openai/agents';
import { z } from 'zod';

const AGENT_ID = 'openai-agents-typescript-example';
const SPAN_ID = 'openai-agents-typescript-example:issue_refund';
const TARGET = 'https://api.example.com/refunds';

function parseEventsOut() {
  const argIndex = process.argv.indexOf('--events-out');
  if (argIndex !== -1 && process.argv[argIndex + 1]) {
    return resolve(process.argv[argIndex + 1]);
  }
  return resolve(
    process.env.ASE_ADAPTER_EVENT_SOURCE ?? 'examples/openai-agents-typescript/events.generated.jsonl',
  );
}

function event(eventType, extra = {}) {
  return {
    protocol_version: 1,
    event_type: eventType,
    event_id: `${eventType}-${Date.now()}-${Math.random().toString(16).slice(2)}`,
    timestamp_ms: Date.now(),
    metadata: {
      adapter_name: 'openai-agents-typescript',
      framework: 'openai-agents-js',
      language: 'typescript',
      transport: 'jsonl-stdio',
    },
    ...extra,
  };
}

function modelResponses() {
  return [
    {
      output: [
        {
          id: 'fc_1',
          type: 'function_call',
          name: 'issue_refund',
          callId: 'call_1',
          status: 'completed',
          arguments: '{"order_id":"ord-001"}',
          providerData: {},
        },
      ],
      usage: new Usage(),
    },
    {
      output: [
        {
          type: 'message',
          role: 'assistant',
          status: 'completed',
          content: [
            {
              type: 'output_text',
              text: 'refunded ord-001',
              providerData: { annotations: [] },
            },
          ],
          providerData: {},
        },
      ],
      usage: new Usage(),
    },
  ];
}

class FakeModel {
  constructor(responses) {
    this.responses = [...responses];
  }

  async getResponse() {
    const response = this.responses.shift();
    if (!response) {
      throw new Error('no fake model response available');
    }
    return response;
  }

  async *getStreamedResponse() {
    throw new Error('streaming is not implemented in this example');
  }
}

function buildAgent(events) {
  const issueRefund = tool({
    name: 'issue_refund',
    description: 'Issue one deterministic refund',
    parameters: z.object({ order_id: z.string() }),
    execute: async ({ order_id }) => {
      events.push(
        event('tool_start', {
          agent_id: AGENT_ID,
          span_id: SPAN_ID,
          tool_kind: 'http_api',
          method: 'POST',
          target: TARGET,
          name: 'issue_refund',
          data: { order_id },
        }),
      );
      events.push(
        event('tool_end', {
          agent_id: AGENT_ID,
          span_id: SPAN_ID,
          tool_kind: 'http_api',
          method: 'POST',
          target: TARGET,
          status: 'passed',
          data: { status_code: 200, order_id },
        }),
      );
      return `refunded ${order_id}`;
    },
  });

  return new Agent({
    name: AGENT_ID,
    instructions: 'Issue a refund with one tool call.',
    model: new FakeModel(modelResponses()),
    tools: [issueRefund],
  });
}

async function main() {
  const eventPath = parseEventsOut();
  const events = [event('agent_start', { agent_id: AGENT_ID, name: AGENT_ID })];
  const agent = buildAgent(events);
  const result = await run(agent, 'refund ord-001');
  events.push(event('agent_end', { agent_id: AGENT_ID, status: 'passed' }));
  await mkdir(dirname(eventPath), { recursive: true });
  await writeFile(eventPath, `${events.map((item) => JSON.stringify(item)).join('\n')}\n`, 'utf-8');
  console.log(result.finalOutput);
  process.exit(0);
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});
